### DISCUSSION TITLE
AttributeError: 'ChatGLMForConditionalGeneration' object has no attribute 'add_adapter'

### DISCUSSION THREAD

**Discussion by lh-kyf** [zai-org/ChatGLM3#498] [Q&A]
chatglm3不支持add_adapter的操作吗？我看llama可以，我使用get_peft_model来实现同样的功能，但是在之后的datacollator上我检查instances都是空的，但是上一步操作还是能看到打印出加载的数据的，有知道是怎么回事的请告诉我，谢谢！

**Comment by zRzRzRzRzRzRzR** [comment]
emm，哪个代码，给个复现，我们支持get_peft_model

**Reply by lh-kyf** [reply]
我自己编写的
基于alpaca的train修改
![image](https://github.com/THUDM/ChatGLM3/assets/116720655/1820f3a9-9a7e-47ad-a063-54a28a9000da)

def make_supervised_data_module(tokenizer: transformers.PreTrainedTokenizer, data_args) -> Dict:
    """Make dataset and collator for supervised fine-tuning."""
    train_dataset = SupervisedDataset(tokenizer=tokenizer, data_path=data_args.data_path)
    data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer)
    return dict(train_dataset=train_dataset, eval_dataset=None, data_collator=data_collator)

def train():
    parser = transformers.HfArgumentParser((ModelArguments, DataArguments, TrainingArguments))
    model_args, data_args, training_args = parser.parse_args_into_dataclasses()

    if model_args.use_flash_attention2:
        model = transformers.AutoModel.from_pretrained(
            model_args.model_name_or_path,
            cache_dir=training_args.cache_dir,
            use_flash_attention_2=True,
            torch_dtype=torch.bfloat16,
            trust_remote_code=True
        )
        print("You are using flash attention 2!\n")
    else:
        model = transformers.AutoModel.from_pretrained(
            model_args.model_name_or_path,
            cache_dir=training_args.cache_dir,
            trust_remote_code=True
        )

    tokenizer = transformers.AutoTokenizer.from_pretrained(
        model_args.model_name_or_path,
        cache_dir=training_args.cache_dir,
        model_max_length=training_args.model_max_length,
        padding_side='left',
        # padding_side="right",
        use_fast=False,
        trust_remote_code=True
    )
   
    if model_args.freeze_layers:
        freeze_model_layers(model, model_args.freeze_layers.split(","))
    if model_args.peft_lora:
        if model_args.lora_config is None:
            raise ValueError("Please specify the path to the PEFT config.")
        lora_config = LoraConfig(**LoraConfig.from_json_file(model_args.lora_config))
        # model.add_adapter(lora_config, adapter_name="adapter_lora")
        model = get_peft_model(model, lora_config)
        print("You are using lora model!\n")

    data_module = make_supervised_data_module(tokenizer=tokenizer, data_args=data_args)
    trainer = Trainer(model=model, tokenizer=tokenizer, args=training_args, **data_module)
    trainer.train()
    trainer.save_state()
    trainer.save_model(output_dir=training_args.output_dir)


if __name__ == "__main__":
    train()

代码在这块就出错了：
@dataclass
class DataCollatorForSeq2Seq(object):
    """Collate examples for supervised fine-tuning."""

    tokenizer: transformers.PreTrainedTokenizer


    def __call__(self, instances: Sequence[Dict]) -> Dict[str, torch.Tensor]:

        print(instances)

        input_ids, labels = tuple([instance[key] for instance in instances] for key in ("input_ids", "labels"))
        input_ids = torch.nn.utils.rnn.pad_sequence(
            input_ids, batch_first=True, padding_value=self.tokenizer.pad_token_id
        )
        labels = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True, padding_value=IGNORE_INDEX)
        return dict(
            input_ids=input_ids,
            labels=labels,
            attention_mask=input_ids.ne(self.tokenizer.pad_token_id),
        )
传入的instance为空

但在
class SupervisedDataset(Dataset):
    """Dataset for supervised fine-tuning."""

    def __init__(self, data_path: str, tokenizer: transformers.PreTrainedTokenizer):
        super(SupervisedDataset, self).__init__()
        logging.warning("Loading data...")
        data_path = data_path.split(",")
        print(data_path)
        list_data_dict = []
        for path in data_path:
            data = jload(path)
            list_data_dict.extend(data)
        print("Data total length: " + str(len(list_data_dict)))
        # print(list_data_dict)
        logging.warning("Formatting inputs...")
        prompt_input, prompt_no_input = PROMPT_DICT["prompt_input"], PROMPT_DICT["prompt_no_input"]
        sources = [
            prompt_input.format_map(example) if example.get("input", "") != "" else prompt_no_input.format_map(example)
            for example in list_data_dict
        ]
        # print(sources)
        targets = [f"{example['output']}{tokenizer.eos_token}" for example in list_data_dict]
        # print(targets)
        logging.warning("Tokenizing inputs... This may take some time...")
        data_dict = preprocess(sources, targets, tokenizer)
        # print(data_dict)

        self.input_ids = data_dict["input_ids"]
        self.labels = data_dict["labels"]


    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, i) -> Dict[str, torch.Tensor]:
        return dict(input_ids=self.input_ids[i], labels=self.labels[i])
还是有数据的

**Comment by lh-kyf** [comment]
我已解决数据传入为空问题，问题在于默认的remove_unused_columns=True上，这里引用我在别的项目上看到的回答：remove_unused_columns=True will remove the keys in the dataset that are not accepted by the models, which is decided by looking at the model forward signature. Your does not have any argument names (it takes ), that's why there is this bug. If you unpack that to use names like , , , the issue will disappear.RewardModel*args*argslabelsinput_idsattention_mask，希望对看到这个问题的人有帮助。当然这个add_adapter我是借鉴别人的，别人也是遇到数据传入为空的问题，但是通过这个方法解决了，不过它的代码是微调llama的，所以这一方面还在研究。
